clippy_lints/src/doc.rs

```diff
@@ -213,13 +215,38 @@ fn lint_for_missing_headers<'a, 'tcx>(
             "unsafe function's docs miss `# Safety` section",
         );
     }
-    if !headers.errors && match_type(cx, return_ty(cx, hir_id), &paths::RESULT) {
-        span_lint(
-            cx,
-            MISSING_ERRORS_DOC,
-            span,
-            "docs for function returning `Result` missing `# Errors` section",
-        );
+    if !headers.errors {
+        if match_type(cx, return_ty(cx, hir_id), &paths::RESULT) {
+            span_lint(
+                cx,
+                MISSING_ERRORS_DOC,
+                span,
+                "docs for function returning `Result` missing `# Errors` section",
+            );
+        } else if let (Some(body_id), Some(future)) = (body_id, get_trait_def_id(cx, &paths::FUTURE)) {
+            let def_id = cx.tcx.hir().body_owner_def_id(body_id);
+            let mir = cx.tcx.optimized_mir(def_id);
+            let ret_ty = mir.return_ty();
+
+            if implements_trait(cx, ret_ty, future, &[]) {
+                use TyKind::*;
+
+                if let Opaque(_, subs) = ret_ty.kind {
+                    if let Some(ty) = subs.types().next() {
+                        if let Generator(_, subs, _) = ty.kind {
+                            if match_type(cx, subs.as_generator().return_ty(def_id, cx.tcx), &paths::RESULT) {
+                                span_lint(
+                                    cx,
+                                    MISSING_ERRORS_DOC,
+                                    span,
+                                    "docs for function returning `Result` missing `# Errors` section",
+                                );
+                            }
+                        }
+                    }
+                }
+            }
```

Comment:
You can use `if_chain` macro to avoid deep nesting.

Reply:
Done.
I don't know why, but the latest nightly broke the clippy testsuite, so I rebased and now I can't even compile locally.
We will see how CI plays out.

Reply:
> I don't know why, but the latest nightly broke the clippy testsuite, so I rebased and now I can't even compile locally.

We build Clippy for latest rustc api so sometimes the build will be broken. You can use specific rustc by using rustup-toolchain-install-master until the fix lands.
More details: https://github.com/rust-lang/rust-clippy/blob/master/CONTRIBUTING.md#fixing-build-failures-caused-by-rust

In this case, #5186 will fix the build.

Reply:
Yeah I just stumbled on this too, thanks! Should I rebase when this lands?

Reply:
We can make sure CI is green so yes, I think.